consensus/src/consensus.rs

```diff
@@ -153,44 +165,94 @@ where
         })
     }
 
+    fn reconfigure(
+        &mut self,
+        new_committee: Committee<PublicKey>,
+    ) -> StoreResult<ConsensusState<PublicKey>> {
+        self.committee = new_committee.clone();
+        self.protocol.update_committee(new_committee)?;
+
+        self.consensus_index = 0;
+
+        let genesis = Certificate::genesis(&self.committee);
+        Ok(ConsensusState::new(genesis, self.metrics.clone()))
+    }
+
     async fn run(&mut self) -> StoreResult<()> {
         // The consensus state (everything else is immutable).
-        let mut state = ConsensusState::new(self.genesis.clone(), self.metrics.clone());
+        let genesis = Certificate::genesis(&self.committee);
+        let mut state = ConsensusState::new(genesis, self.metrics.clone());
 
         // Listen to incoming certificates.
-        while let Some(certificate) = self.rx_primary.recv().await {
-            let sequence =
-                self.protocol
-                    .process_certificate(&mut state, self.consensus_index, certificate)?;
-
-            // Update the consensus index.
-            self.consensus_index += sequence.len() as u64;
-
-            // Output the sequence in the right order.
-            for output in sequence {
-                let certificate = &output.certificate;
-                #[cfg(not(feature = "benchmark"))]
-                if output.consensus_index % 5_000 == 0 {
-                    tracing::debug!("Committed {}", certificate.header);
-                }
+        loop {
+            tokio::select! {
+                Some(certificate) = self.rx_primary.recv() => {
+                    // If the core already moved to the next epoch we should pull the next
+                    // committee as well.
+                    match certificate.epoch().cmp(&self.committee.epoch()) {
+                        Ordering::Greater => {
+                            let message = self.rx_reconfigure.borrow_and_update().clone();
+                            match message  {
+                                ReconfigureNotification::NewCommittee(new_committee) => {
+                                    state = self.reconfigure(new_committee)?;
+                                },
+                                ReconfigureNotification::Shutdown => return Ok(()),
+                            }
+                        }
+                        Ordering::Less => {
+                            // We already updated committee but the core is slow.
```

Comment:
Agreed, it might be spammy - well for a short period of time under healthy circumstances - but especially until we gain good trust on the stability of our nodes I would err on adding more logging rather less (doesn't have to be `info` - it could be a `debug`). For this specific part I would probably like to know if for some reason this component is dropping certificates and not doing any processing. We might be looking situations where the node is not making any progress and I would like to answer the question "why is my consensus component is not making progress?"

Reply:
fixed